### PR TITLE
executor: make projection executor unparallel for insert/update/delete

### LINKED ISSUE #26832: TiDB panic index out of range [4294967295] after long time tiup bench run
### 1. Minimal reproduce step (Required)

run command `tiup bench tpcc --warehouses 10 --parts 10 run -T 128 -p "xxxxx" -D bench -U bencher`

### 2. What did you expect to see? (Required)

`tiup bench` runs smoothly and works fine.

### 3. What did you see instead (Required)

TiDB is down and stderr outputs this

```plain
panic: runtime error: index out of range [4294967295] with length 1

goroutine 114089996 [running]:
github.com/pingcap/tidb/store/tikv/unionstore.(*nodeAllocator).getNode(...)
        /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tidb/store/tikv/unionstore/memdb_arena.go:192
github.com/pingcap/tidb/store/tikv/unionstore.(*MemDB).getNode(...)
        /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tidb/store/tikv/unionstore/memdb.go:733
github.com/pingcap/tidb/store/tikv/unionstore.memdbNodeAddr.getRight(...)
        /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tidb/store/tikv/unionstore/memdb.go:765
github.com/pingcap/tidb/store/tikv/unionstore.(*MemDB).traverse(0xc048215c70, 0xc074bfaf90, 0x13, 0x13, 0x0, 0x60, 0x7fbc4cc25a48)
        /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tidb/store/tikv/unionstore/memdb.go:354 +0xd31
github.com/pingcap/tidb/store/tikv/unionstore.(*MemDB).GetFlags(0xc048215c70, 0xc074bfaf90, 0x13, 0x13, 0x3569680, 0x3b9f101, 0xc074a0bb60)
        /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tidb/store/tikv/unionstore/memdb.go:202 +0x52
github.com/pingcap/tidb/store/tikv/unionstore.(*KVUnionStore).HasPresumeKeyNotExists(0xc0005908b8, 0xc074bfaf90, 0x13, 0x13, 0x13)
        /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tidb/store/tikv/unionstore/union_store.go:120 +0x4c
github.com/pingcap/tidb/store/tikv.actionPessimisticLock.handleSingleBatch(0xc074f003f0, 0xc0479f89a0, 0xc010b6c700, 0x2465, 0x5, 0x1bd, 0x3c05678, 0xc074c0cfc0, 0x0, 0x112fee5, ...)
        /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tidb/store/tikv/pessimistic.go:79 +0x1c7
github.com/pingcap/tidb/store/tikv.(*batchExecutor).startWorker.func1(0xc012715dc0, 0xc074c0d0e0, 0x2465, 0x5, 0x1bd, 0x3c05678, 0xc074c0cfc0, 0x0)
        /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tidb/store/tikv/2pc.go:1641 +0x187
created by github.com/pingcap/tidb/store/tikv.(*batchExecutor).startWorker
        /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tidb/store/tikv/2pc.go:1624 +0x19e
```

### 4. What is your TiDB version? (Required)

```
Release Version: v5.1.0
Edition: Community
Git Commit Hash: 8acd5c88471cb7b4d4c4a8ed73b4d53d6833f13e
Git Branch: heads/refs/tags/v5.1.0
UTC Build Time: 2021-06-24 07:10:32
GoVersion: go1.16.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close #26832

Problem Summary:

The memdb is not thread-safe, and that's the root cause of issue 26832

### What is changed and how it works?

In insert/update/delete statement, do not use the parallel projection, choose the unparallel one intead.

This can avoid the concurrent visit of the memdb in issue 26832.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

I try to add some unit test code like this and run it with `-race` flag, but can't reproduce.

```
+func (s *testSuite8) TestIssues26832DataRace(c *C) {
+       tk := testkit.NewTestKit(c, s.store)
+       tk.MustExec("use test")
+       tk.MustExec("drop table if exists t")
+       tk.MustExec("create table t (a int unique)")
+       tk.MustExec(`CREATE TABLE IF NOT EXISTS order_line (
+               ol_o_id INT NOT NULL,
+               ol_d_id INT NOT NULL,
+               ol_w_id INT NOT NULL,
+               ol_i_id INT NOT NULL,
+               ol_delivery_d INT,
+               PRIMARY KEY(ol_w_id, ol_d_id, ol_o_id)) partition by hash(ol_w_id) partitions 10`)
+       for i:=0; i<1000; i++ {
+               tk.MustExec("insert into order_line value (?, ?, ?, ?, ?)", i, i, i, i, i)
+       }
+
+       for i:=0; i<1000; i++ {
+               v := rand.Intn(1000)
+               tk.MustExec(`UPDATE order_line SET ol_delivery_d = ol_delivery_d + 1 WHERE (ol_w_id, ol_d_id, ol_o_id) IN (
+       (?,?,?),(?,?,?),(?,?,?),(?,?,?),(?,?,?),(?,?,?),(?,?,?),(?,?,?),(?,?,?),(?,?,?))`,
+                       v,v,v,
+                       v,v,v,
+                       v,v,v,
+                       v,v,v,
+                       v,v,v,
+                       v,v,v,
+                       v,v,v,
+                       v,v,v,
+                       v,v,v,
+                       v,v,v)
+       }
+}
```

and then, I try the TPCC on my local machine, and with no luck ... it doesn't reproduce

```
tiup bench tpcc --warehouses 10 --parts 10 prepare -T 16 -D bench
tiup bench tpcc --warehouses 10 --parts 10 run -T 16 -D bench
```

This change is simple enough, so maybe we can merge it first and see whether it reproduce in the future.

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

It should not impact the write performance in most cases, because the insert/update/delete is unparalleled
and be the bottleneck, the change of projection does't change this fact.

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by tiancaiamao** [comment]
/run-check_dev_2

Unstable test https://github.com/pingcap/tidb/issues/30303

**Comment by cfzjywxk** [comment]
Or could we move the `selectLockExec` executor at the top above the projection?

**Comment by tiancaiamao** [comment]
> Or could we move the `selectLockExec` executor at the top above the projection?

`selectLockExec` need some extra column related to the partition column... and if those columns are not used by the parent executor, they need to be pruned, so projection -> selectLockExec is inevitable @cfzjywxk

File: executor/builder.go [hunk 1]
```diff
@@ -1420,6 +1422,12 @@ func (b *executorBuilder) buildProjection(v *plannercore.PhysicalProjection) Exe
 	if int64(v.StatsCount()) < int64(b.ctx.GetSessionVars().MaxChunkSize) {
 		e.numWorkers = 0
 	}
+
+	// Use un-parallel projection for write query to avoid concurrent on memdb.
+	// See also https://github.com/pingcap/tidb/issues/26832
+	if b.inUpdateStmt || b.inDeleteStmt || b.inInsertStmt {
```
Comment (coocood) [inline executor/builder.go:1428]: Maybe we also need to handle `SELECT FOR UPDATE`

**Comment by XuHuaiyu** [comment]
Do we need to cherrypick this commit?

File: executor/builder.go (hunk 1 above)
Comment (tiancaiamao) [reply executor/builder.go]: Done @coocood

**Review by coocood — APPROVED** [review]

**Review by cfzjywxk — APPROVED** [review]
We may need to do some verify tests in the test environment, I think @ChenPeng2013 could help with this.

**Pull request merged**